compiler/transforms/frontend/preprocess_mlperf_tiny.py

```diff
@@ -0,0 +1,312 @@
+
+
+class DropOldFunction(RewritePattern):
+    """
+    After the function `main` has been inlined in the function
+    `run_network`, the main function can be dropped
+    """
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: func.FuncOp, rewriter: PatternRewriter):
+        ## delete old main function which has been inlined
+        if op.sym_name.data == "main":
+            rewriter.erase_matched_op()
+
+
+class AllocToGlobal(RewritePattern):
+    """
+    Convert all static allocs to empty statically scheduled memref globals
+    """
+
+    # keep track of index to give memref globals a name
+    index: int = 0
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: memref.Alloc, rewriter: PatternRewriter):
+        # get module op
+        module_op = op
+        while not isinstance(module_op, builtin.ModuleOp):
+            assert module_op is not None
+            module_op = module_op.parent_op()
+
+        symbol_table = module_op.get_trait(SymbolTable)
+        assert symbol_table
+
+        # create global symbol name
+        global_sym_name = "_static_const_" + str(self.index)
+        self.index = self.index + 1
+
+        # check if it is not in the symbol table
+        assert SymbolTable.lookup_symbol(module_op, global_sym_name) is None
+
+        # create global
+        memref_global = memref.Global.get(
+            builtin.StringAttr(global_sym_name),
+            op.results[0].type,
+            initial_value=builtin.UnitAttr(),
+        )
+        SymbolTable.insert_or_update(module_op, memref_global)
+
+        # remove all the deallocs
+        deallocs = [
+            user_op.operation
+            for user_op in op.results[0].uses
+            if isinstance(user_op.operation, memref.Dealloc)
+        ]
+        for dealloc in deallocs:
+            rewriter.erase_op(dealloc)
+
+        # replace op by get global
+        memref_get = memref.GetGlobal(global_sym_name, op.results[0].type)
+
+        rewriter.replace_matched_op(memref_get)
+
+
+class RemoveZeroInits(RewritePattern):
+    """
+    These ops initialize memory to zero before doing a matmul.
+    With the gemm accelerator we use, this is not necessary.
+    """
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: linalg.FillOp, rewriter: PatternRewriter):
+        # delete fill ops that fill with zero, the accelerator will init from zero by default
+        if len(op.inputs) != 1:
+            return
+        if not isinstance((const_input := op.inputs[0]), OpResult):
+            return
+        if not isinstance((const_op := const_input.op), arith.Constant):
+            return
+        if not isinstance((intattr := const_op.value), builtin.IntegerAttr):
+            return
+        if intattr.value.data != 0:
+            return
+
+        # erase the op
+        rewriter._replace_all_uses_with(op.results[0], op.outputs[0])
+        rewriter.erase_matched_op()
```

Comment:
I guess the difference w.r.t `InsertMemoryClears` is that this has to do with initialization, whereas that has to do more with freeing/allocation?

Reply:
Yes!